pageserver/src/walreceiver.rs

```diff
@@ -64,375 +80,1133 @@ thread_local! {
+            // The wait is only initiated when no data (or a "channel closed" data) is received from the loop, ending with the break flow return.
+            // While waiting, more broker events are expected to be retrieved from etcd (currently, every safekeeper posts ~1 message/second).
+            // The timebomb ensures that we don't get stuck for too long on any of the WAL/etcd event polling, rather restarting the subscription entirely.
+            //
+            // We cannot return here eagerly on no WAL task data, since the result will get selected to early, not allowing etcd tasks to be polled properly.
+            // We cannot move etcd tasks above this select, since they are very frequent to finish and WAL events might get ignored.
+            // We need WAL events to periodically update the external data, so we cannot simply await the task result on the handler here.
+            wal_receiver_poll_result = wal_connection_manager.poll_connection_event_or_cancel() => match wal_receiver_poll_result {
+                ControlFlow::Break(()) => break,
+                ControlFlow::Continue(()) => {},
+            },
+            // finally, if no other tasks are completed, get another broker update and possibly reconnect
+            updates = broker_subscription.fetch_data() => match updates {
+                Some(mut all_timeline_updates) => {
+                    if let Some(subscribed_timeline_updates) = all_timeline_updates.remove(&id) {
+                        match wal_connection_manager.select_connection_candidate(subscribed_timeline_updates) {
+                            Some(candidate) => {
+                                info!("Switching to different safekeeper {} for timeline {id}, reason: {:?}", candidate.safekeeper_id, candidate.reason);
+                                wal_connection_manager.change_connection(candidate.safekeeper_id, candidate.wal_producer_connstr).await;
+                            },
+                            None => {}
+                        }
+                    }
+                },
+                None => {
+                    info!("Subscription source end was dropped, no more updates are possible, shutting down");
+                    break;
+                },
+            },
         }
-    }) {
-        let replication_message = replication_message?;
-        let status_update = match replication_message {
-            ReplicationMessage::XLogData(xlog_data) => {
-                // Pass the WAL data to the decoder, and see if we can decode
-                // more records as a result.
-                let data = xlog_data.data();
-                let startlsn = Lsn::from(xlog_data.wal_start());
-                let endlsn = startlsn + data.len() as u64;
-
-                trace!("received XLogData between {} and {}", startlsn, endlsn);
+    }
 
-                waldecoder.feed_bytes(data);
+    info!("Waiting for the current connection to close");
```

Comment:
BTW, closing connection here means even on short etcd outage we immediately lose current walreceiver connection. We could keep it.